tests/Unit/Parallel/Test_SectionReductions.cpp

```diff
@@ -0,0 +1,248 @@
+          array_index, count, is_even ? "even" : "odd");
+      SPECTRE_PARALLEL_REQUIRE(count == (is_even ? 3 : 2));
+    } else {
+      const bool is_first_element = section_id;
+      Parallel::printf(
+          "Element %d received reduction: Counted %zu element in "
+          "'IsFirstElement' section.\n",
+          array_index, count);
+      SPECTRE_PARALLEL_REQUIRE(is_first_element);
+      SPECTRE_PARALLEL_REQUIRE(count == 1);
+    }
+  }
+};
+
+template <typename Metavariables>
+struct ArrayComponent {
+  using chare_type = Parallel::Algorithms::Array;
+  using metavariables = Metavariables;
+  using array_index = int;
+
+  using phase_dependent_action_list = tmpl::list<
+      Parallel::PhaseActions<typename Metavariables::Phase,
+                             Metavariables::Phase::Initialization,
+                             tmpl::list<>>,
+      Parallel::PhaseActions<
+          typename Metavariables::Phase, Metavariables::Phase::TestReductions,
+          tmpl::list<Count<EvenOrOddTag>,
+                     // Test performing a reduction over another section
+                     Count<IsFirstElementTag>,
+                     // Test performing multiple reductions asynchronously
+                     Count<EvenOrOddTag>, Count<EvenOrOddTag>,
+                     Count<IsFirstElementTag>,
+                     Parallel::Actions::TerminatePhase>>>;
+
+  // [sections_example]
+  using array_allocation_tags = tmpl::list<
+      // The section proxy will be stored in each element's DataBox in this tag
+      // for convenient access
+      Parallel::Tags::Section<ArrayComponent, EvenOrOddTag>,
+      Parallel::Tags::Section<ArrayComponent, IsFirstElementTag>>;
+  using initialization_tags = Parallel::get_initialization_tags<
+      Parallel::get_initialization_actions_list<phase_dependent_action_list>,
+      array_allocation_tags>;
+
+  static void allocate_array(
+      Parallel::CProxy_GlobalCache<Metavariables>& global_cache,
+      tuples::tagged_tuple_from_typelist<initialization_tags>
+          initialization_items) noexcept {
+    auto& local_cache = *(global_cache.ckLocalBranch());
+    auto& array_proxy =
+        Parallel::get_parallel_component<ArrayComponent>(local_cache);
+
+    // Create sections from element IDs (the corresponding elements will be
+    // created below)
+    const size_t num_elements = 5;
+    std::vector<CkArrayIndex> even_elements{};
+    std::vector<CkArrayIndex> odd_elements{};
+    for (size_t i = 0; i < num_elements; ++i) {
+      (i % 2 == 0 ? even_elements : odd_elements)
+          .push_back(Parallel::ArrayIndex<int>(static_cast<int>(i)));
+    }
+    std::vector<CkArrayIndex> first_element{};
+    first_element.push_back(Parallel::ArrayIndex<int>(0));
+    using EvenOrOddSection = Parallel::Section<ArrayComponent, EvenOrOddTag>;
+    const EvenOrOddSection even_section{
+        EvenOrOdd::Even, EvenOrOddSection::cproxy_section::ckNew(
```

Comment:
This appears to violate the requirement from the charm++ documentation that the section proxies be created only in entry methods after all of the elements in question have been inserted.
I admit this appears to _work_ on a single thread, but it would make me feel better if
a) you could test this on multiple threads on your own system, and if it fails add appropriate code-comment warnings, and choose something else for the example. If it doesn't fail, then confirm on a larger case, and if that works fine, then just code-commenting the charm restriction that is violated seems fine. Of course, I may be mistaken, and welcome corrections to my understanding.
b) at least make the proxy construction happen after `doneInserting` if it is possible.

Reply:
Oh I wasn't aware of the charm docs posing that restriction. Are you referring to the warning box here: https://charm.readthedocs.io/en/latest/charm++/manual.html#creation-order-restrictions? I must admit that I didn't quite understand what they mean with "post-constructor entry methods". I'm guessing they mean the array must have been created, and then sections should be created in entry methods of the array (not of the array elements)? I believe that restriction is fulfilled here. It would seem wrong to create sections in entry methods of the array _elements_.

That said, it could be nicer to create the sections after elements have been inserted nonetheless. I'll take a look. To your point a): It seems to work on multiple threads and nodes, I've been running the elliptic solver with this code for a while :)

Reply:
Yeah, that's what I was referring to -- The wording made me concerned that both the creation (via insertion) of new elements and the allocation to sections each required a message to be sent around, and that charm wouldn't guarantee the ordering of those messages, so if both were done without the application (i.e. us) ensuring the ordering that things might break.
This would be analogous to some of the restrictions during balancing, that migration cannot happen before `doneInserting` is called. But if it works consistently, I guess it's probably okay? Best I think to at least make reference to the possible problem in case it shows up in some future charm change and baffles someone who implements the section creation similarly.

Reply:
So I think the _array_ must have been created before _sections_ of the array, but the _elements_ of the array can be created later. In other words, `ckNew` must be called for the array component on the particular PE before `ckNew` is called for the section. That is always guaranteed when creating the sections in `allocate_array`. This is how I read the warning in the Charm++ docs, where they discuss group and array creation order.

I looked at creating sections after element insertion, but the issue with that is that the section proxy has to be sent to the elements asynchronously instead of just copying it into their DataBox right at creation. That's certainly possible, but it adds some complexity that may not be necessary. It's not super easy to send it as a message because the message size needs to be determined somehow (I think -- this is related to variable-sized messages, which we'll probably have to figure out for perfect-forwarding at some point). For now I added a warning to the `Section` docs that creation order _might_ be relevant, as you suggested.

Reply:
That sounds like a reasonable solution for the time being

Reply:
I've asked the Charm folks for clarification on this and will forward anything I learn :)